### PR TITLE
Support macOS (High Sierra)

### PR DESCRIPTION
**Pull request by wooksong**
https://github.com/myungjoo/SSAT/pull/12 is required.

This PR concerns macOS (High Sierra) support.

- Update the meson build script to make NNStreamer be built on macOS
- Support .dylib as a file extension of shared object in the macOS envirionment
- Let unit test cases for common and tensor_sink work on macOS 

### How to test this PR

1. Development environment
```bash
$ uname -a
Darwin MacbookAirAtHome 17.7.0 Darwin Kernel Version 17.7.0: Sun Jun  2 20:31:42 PDT 2019; root:xnu-4570.71.46~1/RELEASE_X86_64 x86_64

$ clang --version
Apple LLVM version 10.0.0 (clang-1000.10.44.4)
Target: x86_64-apple-darwin17.7.0
Thread model: posix
InstalledDir: /Library/Developer/CommandLineTools/usr/bin
```

2. How to build
- [brew](https://brew.sh) is required. 
```bash
$ /usr/bin/ruby -e "$(curl -fsSL https://raw.githubusercontent.com/Homebrew/install/master/install)"
```
- Install packages for build dependency
```bash
$ brew install cask
$ brew update
$ brew install meson pkg-config cmake libffi glib gstreamer gst-plugins-base gst-plugins-good
```
- Build
```bash
$ meson build -Denable-tensorflow=false -Denable-tensorflow-lite=false -Denable-pytorch=false -Denable-caffe2=false
$ ninja -C build
```

3. How to test
- Install googletest
```bash
$ git clone https://github.com/google/googletest
$ cd googletest
$ mkdir build
$ cd build
$ cmake ..
$ make
$ make install
```
- Build nnstreamer again
```bash
$ cd $NNS_DIR
$ rm -rf build
$ meson build -Denable-tensorflow=false -Denable-tensorflow-lite=false -Denable-pytorch=false -Denable-caffe2=false
$ ninja -C build
```

- Set environment variables for test
```bash
$ cd $NNS_DIR
$ pushd build
$ export GST_PLUGIN_PATH=$(pwd)/gst/nnstreamer
$ export NNSTREAMER_CONF=$(pwd)/nnstreamer-test.ini
$ export NNSTREAMER_FILTERS=$(pwd)/ext/nnstreamer/tensor_filter
$ export NNSTREAMER_DECODERS=$(pwd)/ext/nnstreamer/tensor_decoder
$ popd
```

- Run googletest-based test cases
```bash
$ ./tests/unittest_common 
...omission...
[==========] 32 tests from 6 test suites ran. (9 ms total)
[  PASSED  ] 32 tests.
```
```bash
$ ./tests/unittest_sink
...omission...
[==========] 69 tests from 2 test suites ran. (34234 ms total)
[  PASSED  ] 69 tests.
```
```bash
$ ./tests/unittest_plugin
...omission...
[==========] 58 tests from 2 test suites ran. (94 ms total)
[  PASSED  ] 58 tests.
```
Signed-off-by: Wook Song <wook16.song@samsung.com>

### CODE REVIEW

**Comment by jaeyun-jung** [comment]
Please consider sub-plugins property in tensor-filter / tensor-decoder. :)
This uses file name (supposed .so) to print sub-plugins list.

```
gst-inspect-1.0 tensor_filter
  sub-plugins         : Registrable sub-plugins list
                        flags: readable
                        String. Default: "custom,tensorflow,python3,tensorflow-lite,pytorch,caffe2"
```
```
gst-inspect-1.0 tensor_decoder

  sub-plugins         : Registrable sub-plugins list
                        flags: readable
                        String. Default: "pose_estimation,bounding_boxes,image_labeling,direct_video"
```

**Comment by wooksong** [comment]
> Please consider sub-plugins property in tensor-filter / tensor-decoder. :)
> This uses file name (supposed .so) to print sub-plugins list.
> 
> ```
> gst-inspect-1.0 tensor_filter
>   sub-plugins         : Registrable sub-plugins list
>                         flags: readable
>                         String. Default: "custom,tensorflow,python3,tensorflow-lite,pytorch,caffe2"
> ```
> 
> ```
> gst-inspect-1.0 tensor_decoder
> 
>   sub-plugins         : Registrable sub-plugins list
>                         flags: readable
>                         String. Default: "pose_estimation,bounding_boxes,image_labeling,direct_video"
> ```

@jaeyun-jung You are right. I will make another PR. FYI,
```bash
$ gst-inspect-1.0 tensor_decoder
...omission...
 sub-plugins         : Registrable sub-plugins list
   flags: readable
   String. Default: "bounding_boxes.dy,pose_estimation.dy,direct_video.dy,image_labeling.dy"
```

File: gst/nnstreamer/nnstreamer.c [hunk 1]
```diff
@@ -57,7 +57,9 @@
 #include "tensor_repo/tensor_reposink.h"
 #include "tensor_repo/tensor_reposrc.h"
 #include "tensor_sink/tensor_sink.h"
+#if !defined(__MACOS__)
 #include "tensor_source/tensor_src_iio.h"
+#endif /* !__MACOS__ */
```
Comment (myungjoo) [inline gst/nnstreamer/nnstreamer.c]: I'd like to recommend ```#if defined(__LINUX???__)```

File: gst/nnstreamer/nnstreamer.c [hunk 2]
```diff
@@ -85,9 +87,11 @@ gst_nnstreamer_init (GstPlugin * plugin)
   NNSTREAMER_INIT (plugin, reposink, REPOSINK);
   NNSTREAMER_INIT (plugin, reposrc, REPOSRC);
   NNSTREAMER_INIT (plugin, sink, SINK);
-  NNSTREAMER_INIT (plugin, src_iio, SRC_IIO);
   NNSTREAMER_INIT (plugin, split, SPLIT);
   NNSTREAMER_INIT (plugin, transform, TRANSFORM);
+#if !defined(__MACOS__)
```
Comment (myungjoo) [inline gst/nnstreamer/nnstreamer.c]: Same here.

**Comment by myungjoo** [comment]
Just a few minor nitpicks... other than that, LGTM.

**Comment by wooksong** [comment]
> > Please consider sub-plugins property in tensor-filter / tensor-decoder. :)
> > This uses file name (supposed .so) to print sub-plugins list.
> > ```
> > gst-inspect-1.0 tensor_filter
> >   sub-plugins         : Registrable sub-plugins list
> >                         flags: readable
> >                         String. Default: "custom,tensorflow,python3,tensorflow-lite,pytorch,caffe2"
> > ```
> > 
> > 
> > ```
> > gst-inspect-1.0 tensor_decoder
> > 
> >   sub-plugins         : Registrable sub-plugins list
> >                         flags: readable
> >                         String. Default: "pose_estimation,bounding_boxes,image_labeling,direct_video"
> > ```
> 
> @jaeyun-jung
> 
> Here are results from macOS. Does it looks OK?
> 
> ```shell
> $ gst-inspect-1.0 tensor_filter
> ...omission...
>   sub-plugins         : Registrable sub-plugins list
>     flags: readable
>     String. Default: "custom"
> 
> $ gst-inspect-1.0 tensor_decoder
> ...omission...
>   sub-plugins         : Registrable sub-plugins list
>     flags: readable
>     String. Default: "bounding_boxes.dy,pose_estimation.dy,direct_video.dy,image_labeling.dy"
> ```

@jaeyun-jung As we discussed, I will submit **another PR** to fix the issue which you gave to me.

**Comment by wooksong** [comment]
## V2
```c
#if !defined(__MACOS__)
```
 is changed to 
```c
#if defined(__gnu_linux__) && !defined(__ANDROID__)
```

File: gst/nnstreamer/nnstreamer.c (hunk 2 above)
Comment (wooksong) [reply gst/nnstreamer/nnstreamer.c]: @myungjoo  Done!

File: gst/nnstreamer/nnstreamer.c (hunk 1 above)
Comment (wooksong) [reply gst/nnstreamer/nnstreamer.c]: @myungjoo  Done!

**Pull request merged**